src/shared/reboot-util.c

```diff
@@ -107,3 +125,63 @@ int shall_restore_state(void) {
 
         return r > 0 ? ret : true;
 }
+
+static int xen_kexec_loaded(void) {
+#if HAVE_XENCTRL
+        size_t size;
+        _cleanup_close_ int privcmd_fd = -1;
+        _cleanup_close_ int buf_fd = -1;
+        void *buffer;
+        xen_kexec_status_t *status;
+        int r;
+
+        size = page_size();;
+        if (sizeof(*status) > size)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "page_size is too small for hypercall");
+
+        privcmd_fd = open("/dev/xen/privcmd", O_RDWR|O_CLOEXEC);
+        if (privcmd_fd == -1)
+                return log_debug_errno(errno, "Cannot access /dev/xen/privcmd: %m");
+
+        buf_fd = open("/dev/xen/hypercall", O_RDWR|O_CLOEXEC);
+        if (buf_fd == -1)
+                return log_debug_errno(errno, "Cannot access /dev/xen/hypercall: %m");
+
+        buffer = mmap(NULL, size, PROT_READ|PROT_WRITE, MAP_SHARED, buf_fd, 0);
+        if (buffer == MAP_FAILED)
+                return log_error_errno(errno, "Cannot allocate buffer for hypercall: %m");
+        memset(buffer, 0, size);
+
+        status = buffer;
+        status->type = KEXEC_TYPE_DEFAULT;
```

Comment:
inset of the memzero and the explicit fiedl initialization here, maybe also use compound initialization here, i.e.:

```c
*(*xen_kexec_status_t) buffer = (xen_kexec_status_t) {
        .type = KEXEC_TYPE_DEFAULT;
};
```
(of course, this would initialize less than the whole page, most likely, but that should be ok, no?)